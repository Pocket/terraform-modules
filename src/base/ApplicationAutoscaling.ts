import { Resource } from 'cdktf';
import { appautoscaling, cloudwatch, iam } from '@cdktf/provider-aws';
const { AppautoscalingPolicy, AppautoscalingTarget } = appautoscaling;
const { CloudwatchMetricAlarm } = cloudwatch;
const { IamRole, IamRolePolicy, DataAwsIamPolicyDocument } = iam;
import { Construct } from 'constructs';

export interface ApplicationAutoscalingProps {
  ecsClusterName: string;
  ecsServiceName: string;
  prefix: string;
  scalableDimension: string;
  scaleInThreshold: number;
  scaleOutThreshold: number;
  stepScaleInAdjustment: number;
  stepScaleOutAdjustment: number;
  tags?: { [key: string]: string };
  targetMaxCapacity: number;
  targetMinCapacity: number;
}

/*
 * Generates an AutoScaling group
 */

export class ApplicationAutoscaling extends Resource {
  constructor(
    scope: Construct,
    name: string,
    config: ApplicationAutoscalingProps
  ) {
    super(scope, name);

    // set up IAM role & policy
    const iamRole = ApplicationAutoscaling.generateIamRole(this, config);
    ApplicationAutoscaling.generateIamRolePolicy(this, config, iamRole);

    // set up autoscaling target & in/out policies
    const autoScalingTarget = ApplicationAutoscaling.generateAutoScalingTarget(
      this,
      config,
      iamRole
    );

    const applicationScaleOut =
      ApplicationAutoscaling.generateAutoSclaingPolicy(
        this,
        config,
        autoScalingTarget,
        'Out'
      );

    const applicationScaleIn = ApplicationAutoscaling.generateAutoSclaingPolicy(
      this,
      config,
      autoScalingTarget,
      'In'
    );

    // set up cloudwatch alarms
    ApplicationAutoscaling.generateCloudwatchMetricAlarm(
      this,
      config,
      'scale_out_alarm',
      `${config.prefix} Service High CPU`,
      'Alarm to add capacity if container CPU is high',
      'GreaterThanThreshold',
      config.scaleOutThreshold,
      applicationScaleOut.arn
    );

    ApplicationAutoscaling.generateCloudwatchMetricAlarm(
      this,
      config,
      'scale_in_alarm',
      `${config.prefix} Service Low CPU`,
      'Alarm to reduce capacity if container CPU is low',
      'LessThanThreshold',
      config.scaleInThreshold,
      applicationScaleIn.arn
    );
  }

  /**
   * Creates an IAM Role
   * @param resource
   * @param config
   * @returns IamRole
   */
  static generateIamRole(
    resource: Resource,
    config: ApplicationAutoscalingProps
  ): iam.IamRole {
    return new IamRole(resource, `autoscaling_role`, {
      name: `${config.prefix}-AutoScalingRole`,
      assumeRolePolicy: new DataAwsIamPolicyDocument(
        resource,
        `autoscaling_assume`,
        {
          statement: [
            {
              effect: 'Allow',
              actions: ['sts:AssumeRole'],
              principals: [
                {
                  identifiers: ['ecs.application-autoscaling.amazonaws.com'],
                  type: 'Service',
                },
              ],
            },
          ],
        }
      ).json,
    });
  }

  /**
   * Creates an IAM Role Policy
   *
   * @param resource
   * @param config
   * @param iamRole
   */
  static generateIamRolePolicy(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    iamRole: iam.IamRole
  ): void {
    new IamRolePolicy(resource, `autoscaling_role_policy`, {
      name: `${config.prefix}-AutoScalingPolicy`,
      role: iamRole.id,
      policy: new DataAwsIamPolicyDocument(resource, `role_policy`, {
        statement: [
          {
            effect: 'Allow',
            actions: [
              'cloudwatch:PutMetricAlarm',
              'cloudwatch:DescribeAlarms',
              'cloudwatch:DeleteAlarms',
            ],
            resources: [
              `arn:aws:cloudwatch:*:*:alarm:/${config.ecsServiceName}*`,
            ],
          },
          {
            effect: 'Allow',
            actions: ['ecs:DescribeServices', 'ecs:UpdateService'],
            resources: [`arn:aws:ecs:*:*:service/${config.ecsServiceName}*`],
          },
        ],
      }).json,
    });
  }

  /**
   * Creates an Auto Scaling Target
   * @param resource
   * @param config
   * @param iamRole
   * @returns AppautoscalingTarget
   */
  static generateAutoScalingTarget(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    iamRole: iam.IamRole
  ): appautoscaling.AppautoscalingTarget {
    return new AppautoscalingTarget(resource, `autoscaling_target`, {
      maxCapacity: config.targetMaxCapacity,
      minCapacity: config.targetMinCapacity,
      resourceId: `service/${config.ecsClusterName}/${config.ecsServiceName}`,
      roleArn: iamRole.arn,
      scalableDimension: 'ecs:service:DesiredCount',
      serviceNamespace: 'ecs',
    });
  }

  /**
   * Creates an Autoscaling Policy
   * @param resource
   * @param config
   * @param target
   * @param type
   * @returns AppautoscalingPolicy
   */
  static generateAutoSclaingPolicy(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    target: appautoscaling.AppautoscalingTarget,
    type: 'In' | 'Out'
  ): appautoscaling.AppautoscalingPolicy {
    let stepAdjustment;

    if (type === 'In') {
      stepAdjustment = [
        {
          metricIntervalUpperBound: '0',
          scalingAdjustment: config.stepScaleInAdjustment,
        },
      ];
    } else {
      stepAdjustment = [
        {
          metricIntervalLowerBound: '0',
          scalingAdjustment: config.stepScaleOutAdjustment,
        },
      ];
    }

    return new AppautoscalingPolicy(
      resource,
      `scale_${type.toLowerCase()}_policy`,
      {
        name: `${config.prefix}-Scale${type}Policy`,
        policyType: 'StepScaling',
        resourceId: target.resourceId,
        scalableDimension: target.scalableDimension,
        serviceNamespace: target.serviceNamespace,

        stepScalingPolicyConfiguration: {
          adjustmentType: `ChangeInCapacity`,
          cooldown: 60,
          metricAggregationType: 'Average',
          stepAdjustment,
        },
        dependsOn: [target],
      }
    );
  }

  /**
   * Creates a Cloudwatch Metric Alarm
   * @param resource
   * @param config
   * @param id
   * @param name
   * @param desc
   * @param operator
   * @param threshold
   * @param arn
   */
  static generateCloudwatchMetricAlarm(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    id: string,
    name: string,
    desc: string,
    operator: string,
    threshold: number,
    arn: string
  ): void {
    new CloudwatchMetricAlarm(resource, id, {
      alarmName: name,
      alarmDescription: desc,
      comparisonOperator: operator,
      evaluationPeriods: 2,
      threshold,
      statistic: 'Average',
      period: 60,
      namespace: 'AWS/ECS',
      metricName: 'CPUUtilization',
      treatMissingData: 'notBreaching',
      dimensions: {
        ClusterName: config.ecsClusterName,
        ServiceName: config.ecsServiceName,
      },
      alarmActions: [arn],
      tags: config.tags,
    });
  }
}
