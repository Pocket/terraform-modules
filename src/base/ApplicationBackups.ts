import { Construct } from 'constructs';
import { TerraformMetaArguments } from 'cdktf';
import {
  BackupPlanRule,
  BackupPlan,
} from '@cdktf/provider-aws/lib/backup-plan';
import {
  BackupSelection,
  BackupSelectionSelectionTag,
} from '@cdktf/provider-aws/lib/backup-selection';
import { BackupVault } from '@cdktf/provider-aws/lib/backup-vault';
import { BackupVaultPolicy } from '@cdktf/provider-aws/lib/backup-vault-policy';

export interface ApplicationBackupProps extends TerraformMetaArguments {
  name: string;
  kmsKeyArn: string;
  prefix: string;
  accountId: string;
  vaultPolicy: string;
  backupPlans: {
    name: string;
    resources: string[];
    rules: Omit<BackupPlanRule, 'targetVaultName'>[];
    selectionTag: BackupSelectionSelectionTag[];
  }[];
  tags?: { [key: string]: string };
}

export class ApplicationBackup extends Construct {
  public backupPlan: BackupPlan;
  public backupSelection: BackupSelection;
  public backupPlanRule: BackupPlanRule;
  private static vault: BackupVault;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationBackupProps,
  ) {
    super(scope, name);

    const vault = new BackupVault(this, 'backup-vault', {
      name: `${config.prefix}-${config.name}`,
      kmsKeyArn: config.kmsKeyArn,
      tags: config.tags,
      provider: config.provider,
    });

    new BackupVaultPolicy(this, 'backup-vault-policy', {
      backupVaultName: vault.name,
      policy: config.vaultPolicy,
      provider: config.provider,
    });

    config.backupPlans.forEach((plan) => {
      const backupPlan = new BackupPlan(this, 'backup-plan', {
        name: plan.name,
        rule: plan.rules.map((rule) => ({
          ...rule,
          targetVaultName: vault.name,
        })),
        tags: config.tags,
        provider: config.provider,
      });

      new BackupSelection(this, 'backup-selection', {
        name: `${config.prefix}-Backup-Selection`,
        planId: backupPlan.id,
        iamRoleArn: `arn:aws:iam::${config.accountId}:role/service-role/AWSBackupDefaultServiceRole`,
        resources: plan.resources,
        selectionTag: plan.selectionTag,
        provider: config.provider,
      });
    });
  }
}
