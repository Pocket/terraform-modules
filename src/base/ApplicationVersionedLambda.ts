import { Fn, Resource } from 'cdktf';
import { Construct } from 'constructs';
import { cloudwatch, iam, lambdafunction, s3 } from '@cdktf/provider-aws';
import {
  DataArchiveFile,
  DataArchiveFileSource,
} from '@cdktf/provider-archive';

export enum LAMBDA_RUNTIMES {
  PYTHON38 = 'python3.8',
  NODEJS12 = 'nodejs12.x',
  NODEJS14 = 'nodejs14.x',
}

export interface ApplicationVersionedLambdaProps {
  name: string;
  description?: string;
  runtime: LAMBDA_RUNTIMES;
  handler: string;
  timeout?: number;
  environment?: { [key: string]: string };
  vpcConfig?: lambdafunction.LambdaFunctionVpcConfig;
  executionPolicyStatements?: iam.DataAwsIamPolicyDocumentStatement[];
  tags?: { [key: string]: string };
  logRetention?: number;
  s3Bucket: string;
  usesCodeDeploy?: boolean;
}

const DEFAULT_TIMEOUT = 5;
const DEFAULT_RETENTION = 14;

export class ApplicationVersionedLambda extends Resource {
  public readonly versionedLambda: lambdafunction.LambdaAlias;
  public lambdaExecutionRole: iam.IamRole;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationVersionedLambdaProps
  ) {
    super(scope, name);

    this.createCodeBucket();
    this.versionedLambda = this.createLambdaFunction();
  }

  private createLambdaFunction() {
    this.lambdaExecutionRole = new iam.IamRole(this, 'execution-role', {
      name: `${this.config.name}-ExecutionRole`,
      assumeRolePolicy: this.getLambdaAssumePolicyDocument(),
    });

    const executionPolicy = new iam.IamPolicy(this, 'execution-policy', {
      name: `${this.config.name}-ExecutionRolePolicy`,
      policy: this.getLambdaExecutionPolicyDocument(),
    });

    new iam.IamRolePolicyAttachment(this, 'execution-role-policy-attachment', {
      role: this.lambdaExecutionRole.name,
      policyArn: executionPolicy.arn,
      dependsOn: [this.lambdaExecutionRole, executionPolicy],
    });

    const defaultLambda = this.getDefaultLambda();
    const lambdaConfig: lambdafunction.LambdaFunctionConfig = {
      functionName: `${this.config.name}-Function`,
      filename: defaultLambda.outputPath,
      handler: this.config.handler,
      runtime: this.config.runtime,
      timeout: this.config.timeout ?? DEFAULT_TIMEOUT,
      sourceCodeHash: defaultLambda.outputBase64Sha256,
      role: this.lambdaExecutionRole.arn,
      vpcConfig: this.config.vpcConfig,
      publish: true,
      lifecycle: {
        ignoreChanges: [
          'filename',
          'source_code_hash',
          this.shouldIgnorePublish() ? 'publish' : '',
        ].filter((v: string) => v),
      },
      tags: this.config.tags,
      environment: this.config.environment
        ? { variables: this.config.environment }
        : undefined,
    };

    const lambda = new lambdafunction.LambdaFunction(
      this,
      'lambda',
      lambdaConfig
    );

    new cloudwatch.CloudwatchLogGroup(this, 'log-group', {
      name: `/aws/lambda/${lambda.functionName}`,
      retentionInDays: this.config.logRetention ?? DEFAULT_RETENTION,
      dependsOn: [lambda],
    });

    return new lambdafunction.LambdaAlias(this, 'alias', {
      functionName: lambda.functionName,
      functionVersion: Fn.element(Fn.split(':', lambda.qualifiedArn), 7),
      name: 'DEPLOYED',
      lifecycle: {
        ignoreChanges: ['function_version'],
      },
      dependsOn: [lambda],
    });
  }

  private shouldIgnorePublish() {
    if (this.config.usesCodeDeploy !== undefined) {
      return this.config.usesCodeDeploy;
    }

    return false;
  }

  private getLambdaAssumePolicyDocument() {
    return new iam.DataAwsIamPolicyDocument(this, 'assume-policy-document', {
      version: '2012-10-17',
      statement: [
        {
          effect: 'Allow',
          actions: ['sts:AssumeRole'],
          principals: [
            {
              identifiers: ['lambda.amazonaws.com', 'edgelambda.amazonaws.com'],
              type: 'Service',
            },
          ],
        },
      ],
    }).json;
  }

  private getLambdaExecutionPolicyDocument() {
    const document = {
      version: '2012-10-17',
      statement: [
        {
          effect: 'Allow',
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents',
            'logs:DescribeLogStreams',
          ],
          resources: ['arn:aws:logs:*:*:*'],
        },
        ...(this.config.executionPolicyStatements ?? []),
      ],
    };

    if (this.config.vpcConfig) {
      document.statement.push({
        effect: 'Allow',
        actions: [
          'ec2:DescribeNetworkInterfaces',
          'ec2:CreateNetworkInterface',
          'ec2:DeleteNetworkInterface',
          'ec2:DescribeInstances',
          'ec2:AttachNetworkInterface',
        ],
        resources: ['*'],
      });
    }

    return new iam.DataAwsIamPolicyDocument(
      this,
      'execution-policy-document',
      document
    ).json;
  }

  private getDefaultLambda() {
    const source = this.getDefaultLambdaSource();
    return new DataArchiveFile(this, 'lambda-default-file', {
      type: 'zip',
      source: [source],
      outputPath: `${source.filename}.zip`,
    });
  }

  private getDefaultLambdaSource(): DataArchiveFileSource {
    const runtime = this.config.runtime.match(/[a-z]*/)[0];
    const handler = this.config.handler.split('.');
    const functionName = handler.pop();
    const functionFilename = handler.join('.');

    let content = `exports.${functionName} = (event, context) => { console.log(event) }`;
    let filename = `${functionFilename}.js`;

    if (runtime === 'python') {
      content = `import json\ndef ${functionName}(event, context):\n\t print(event)\n\t return {'statusCode': 200, 'headers': {'dance': 'party'}, 'body': json.dumps({'electric': 'boogaloo'}), 'isBase64Encoded': False}`;
      filename = `${functionFilename}.py`;
    }

    return {
      content,
      filename,
    };
  }

  private createCodeBucket() {
    const codeBucket = new s3.S3Bucket(this, 'code-bucket', {
      bucket: this.config.s3Bucket,
      acl: 'private',
      tags: this.config.tags,
      forceDestroy: true,
    });

    new s3.S3BucketPublicAccessBlock(this, `code-bucket-public-access-block`, {
      bucket: codeBucket.id,
      blockPublicAcls: true,
      blockPublicPolicy: true,
    });
  }
}
