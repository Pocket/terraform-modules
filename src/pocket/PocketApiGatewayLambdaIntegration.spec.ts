import { TerraformStack, Testing } from 'cdktf';
import { LAMBDA_RUNTIMES } from '../base/ApplicationVersionedLambda';
import {
  PocketApiGateway,
  PocketApiGatewayProps,
} from './PocketApiGatewayLambdaIntegration';

const config: PocketApiGatewayProps = {
  name: 'test-api-lambda',
  region: 'us-east-1',
  accountId: 'abc123',
  routes: [
    {
      path: 'endpoint',
      method: 'POST',
      eventHandler: {
        name: 'lambda-endpoint',
        lambda: {
          runtime: LAMBDA_RUNTIMES.PYTHON38,
          handler: 'index.handler',
        },
      },
    },
  ],
};

test('renders an api gateway with a lambda integration', () => {
  const synthed = Testing.synthScope((stack) => {
    new PocketApiGateway(stack, 'test-api-lambda', {
      ...config,
    });
  });
  expect(synthed).toMatchSnapshot();
});
