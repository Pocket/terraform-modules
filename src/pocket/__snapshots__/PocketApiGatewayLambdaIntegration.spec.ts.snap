// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PocketApiGatewayLambdaIntegration renders an api gateway with a lambda integration 1`] = `
"{
  \\"data\\": {
    \\"archive_file\\": {
      \\"test-api-lambda_endpoint-lambda_lambda-default-file_8FD40136\\": {
        \\"output_path\\": \\"index.py.zip\\",
        \\"source\\": [
          {
            \\"content\\": \\"import json\\\\ndef handler(event, context):\\\\n\\\\t print(event)\\\\n\\\\t return {'statusCode': 200, 'headers': {'dance': 'party'}, 'body': json.dumps({'electric': 'boogaloo'}), 'isBase64Encoded': False}\\",
            \\"filename\\": \\"index.py\\"
          }
        ],
        \\"type\\": \\"zip\\"
      }
    },
    \\"aws_iam_policy_document\\": {
      \\"test-api-lambda_endpoint-lambda_assume-policy-document_345E1069\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"lambda.amazonaws.com\\",
                  \\"edgelambda.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      },
      \\"test-api-lambda_endpoint-lambda_execution-policy-document_4437ECD0\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"logs:CreateLogGroup\\",
              \\"logs:CreateLogStream\\",
              \\"logs:PutLogEvents\\",
              \\"logs:DescribeLogStreams\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"arn:aws:logs:*:*:*\\"
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      }
    },
    \\"aws_route53_zone\\": {
      \\"test-api-lambda_base-dns_main_hosted_zone_5F97294A\\": {
        \\"name\\": \\"getpocket.dev\\"
      }
    }
  },
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"test-api-lambda_api-gateway-certificate_8D77B940\\": {
        \\"domain_name\\": \\"exampleapi.getpocket.dev\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"validation_method\\": \\"DNS\\"
      }
    },
    \\"aws_acm_certificate_validation\\": {
      \\"test-api-lambda_api-gateway-certificate_certificate_validation_6C09E593\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.test-api-lambda_api-gateway-certificate_8D77B940.arn}\\",
        \\"depends_on\\": [
          \\"aws_route53_record.test-api-lambda_api-gateway-certificate_certificate_record_BED7C8FC\\",
          \\"aws_acm_certificate.test-api-lambda_api-gateway-certificate_8D77B940\\"
        ],
        \\"validation_record_fqdns\\": [
          \\"\${aws_route53_record.test-api-lambda_api-gateway-certificate_certificate_record_BED7C8FC.fqdn}\\"
        ]
      }
    },
    \\"aws_api_gateway_base_path_mapping\\": {
      \\"test-api-lambda_api-gateway-base-path-mapping_CC4D8627\\": {
        \\"api_id\\": \\"\${aws_api_gateway_rest_api.api-gateway-rest.id}\\",
        \\"base_path\\": \\"fxaProxy\\",
        \\"domain_name\\": \\"\${aws_api_gateway_domain_name.test-api-lambda_api-gateway-domain-name_6DBA5B05.domain_name}\\",
        \\"stage_name\\": \\"\${aws_api_gateway_stage.api-gateway-stage.stage_name}\\"
      }
    },
    \\"aws_api_gateway_deployment\\": {
      \\"api-gateway-deployment\\": {
        \\"depends_on\\": [
          \\"aws_api_gateway_integration.test-api-lambda_endpoint-integration_8D80A8F8\\",
          \\"aws_api_gateway_method.test-api-lambda_endpoint-method_B2F73D1A\\",
          \\"aws_api_gateway_resource.test-api-lambda_endpoint_02B05EB1\\",
          \\"aws_lambda_alias.test-api-lambda_endpoint-lambda_alias_A14F4FF8\\"
        ],
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.api-gateway-rest.id}\\",
        \\"triggers\\": {
          \\"deployedAt\\": \\"1637693316456\\",
          \\"redeployment\\": \\"\${sha1(jsonencode({resources = [aws_api_gateway_integration.test-api-lambda_endpoint-integration_8D80A8F8.id, aws_api_gateway_method.test-api-lambda_endpoint-method_B2F73D1A.id, aws_api_gateway_resource.test-api-lambda_endpoint_02B05EB1.id, aws_lambda_alias.test-api-lambda_endpoint-lambda_alias_A14F4FF8.id]}))}\\"
        }
      }
    },
    \\"aws_api_gateway_domain_name\\": {
      \\"test-api-lambda_api-gateway-domain-name_6DBA5B05\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.test-api-lambda_api-gateway-certificate_8D77B940.arn}\\",
        \\"depends_on\\": [
          \\"aws_acm_certificate_validation.test-api-lambda_api-gateway-certificate_certificate_validation_6C09E593\\"
        ],
        \\"domain_name\\": \\"exampleapi.getpocket.dev\\"
      }
    },
    \\"aws_api_gateway_integration\\": {
      \\"test-api-lambda_endpoint-integration_8D80A8F8\\": {
        \\"http_method\\": \\"POST\\",
        \\"integration_http_method\\": \\"POST\\",
        \\"resource_id\\": \\"\${aws_api_gateway_resource.test-api-lambda_endpoint_02B05EB1.id}\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.api-gateway-rest.id}\\",
        \\"type\\": \\"AWS_PROXY\\",
        \\"uri\\": \\"\${aws_lambda_alias.test-api-lambda_endpoint-lambda_alias_A14F4FF8.invoke_arn}\\"
      }
    },
    \\"aws_api_gateway_method\\": {
      \\"test-api-lambda_endpoint-method_B2F73D1A\\": {
        \\"authorization\\": \\"NONE\\",
        \\"http_method\\": \\"POST\\",
        \\"resource_id\\": \\"\${aws_api_gateway_resource.test-api-lambda_endpoint_02B05EB1.id}\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.api-gateway-rest.id}\\"
      }
    },
    \\"aws_api_gateway_resource\\": {
      \\"test-api-lambda_endpoint_02B05EB1\\": {
        \\"parent_id\\": \\"\${aws_api_gateway_rest_api.api-gateway-rest.root_resource_id}\\",
        \\"path_part\\": \\"endpoint\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.api-gateway-rest.id}\\"
      }
    },
    \\"aws_api_gateway_rest_api\\": {
      \\"api-gateway-rest\\": {
        \\"name\\": \\"test-api-lambda\\"
      }
    },
    \\"aws_api_gateway_stage\\": {
      \\"api-gateway-stage\\": {
        \\"deployment_id\\": \\"\${aws_api_gateway_deployment.api-gateway-deployment.id}\\",
        \\"rest_api_id\\": \\"\${aws_api_gateway_rest_api.api-gateway-rest.id}\\",
        \\"stage_name\\": \\"test\\"
      }
    },
    \\"aws_cloudwatch_log_group\\": {
      \\"test-api-lambda_endpoint-lambda_log-group_E143678F\\": {
        \\"depends_on\\": [
          \\"aws_lambda_function.test-api-lambda_endpoint-lambda_3FA91964\\"
        ],
        \\"name\\": \\"/aws/lambda/\${aws_lambda_function.test-api-lambda_endpoint-lambda_3FA91964.function_name}\\",
        \\"retention_in_days\\": 14
      }
    },
    \\"aws_iam_policy\\": {
      \\"test-api-lambda_endpoint-lambda_execution-policy_6847E583\\": {
        \\"name\\": \\"lambda-endpoint-ExecutionRolePolicy\\",
        \\"policy\\": \\"\${data.aws_iam_policy_document.test-api-lambda_endpoint-lambda_execution-policy-document_4437ECD0.json}\\"
      }
    },
    \\"aws_iam_role\\": {
      \\"test-api-lambda_endpoint-lambda_execution-role_46759F27\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test-api-lambda_endpoint-lambda_assume-policy-document_345E1069.json}\\",
        \\"name\\": \\"lambda-endpoint-ExecutionRole\\"
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test-api-lambda_endpoint-lambda_execution-role-policy-attachment_5BF51F3C\\": {
        \\"depends_on\\": [
          \\"aws_iam_role.test-api-lambda_endpoint-lambda_execution-role_46759F27\\",
          \\"aws_iam_policy.test-api-lambda_endpoint-lambda_execution-policy_6847E583\\"
        ],
        \\"policy_arn\\": \\"\${aws_iam_policy.test-api-lambda_endpoint-lambda_execution-policy_6847E583.arn}\\",
        \\"role\\": \\"\${aws_iam_role.test-api-lambda_endpoint-lambda_execution-role_46759F27.name}\\"
      }
    },
    \\"aws_lambda_alias\\": {
      \\"test-api-lambda_endpoint-lambda_alias_A14F4FF8\\": {
        \\"depends_on\\": [
          \\"aws_lambda_function.test-api-lambda_endpoint-lambda_3FA91964\\"
        ],
        \\"function_name\\": \\"\${aws_lambda_function.test-api-lambda_endpoint-lambda_3FA91964.function_name}\\",
        \\"function_version\\": \\"\${element(split(\\\\\\":\\\\\\", aws_lambda_function.test-api-lambda_endpoint-lambda_3FA91964.qualified_arn), 7)}\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"function_version\\"
          ]
        },
        \\"name\\": \\"DEPLOYED\\"
      }
    },
    \\"aws_lambda_function\\": {
      \\"test-api-lambda_endpoint-lambda_3FA91964\\": {
        \\"filename\\": \\"\${data.archive_file.test-api-lambda_endpoint-lambda_lambda-default-file_8FD40136.output_path}\\",
        \\"function_name\\": \\"lambda-endpoint-Function\\",
        \\"handler\\": \\"index.handler\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"filename\\",
            \\"source_code_hash\\"
          ]
        },
        \\"memory_size\\": 128,
        \\"publish\\": true,
        \\"reserved_concurrent_executions\\": -1,
        \\"role\\": \\"\${aws_iam_role.test-api-lambda_endpoint-lambda_execution-role_46759F27.arn}\\",
        \\"runtime\\": \\"python3.8\\",
        \\"source_code_hash\\": \\"\${data.archive_file.test-api-lambda_endpoint-lambda_lambda-default-file_8FD40136.output_base64sha256}\\",
        \\"timeout\\": 5
      }
    },
    \\"aws_lambda_permission\\": {
      \\"test-api-lambda_test-api-lambda_endpoint-lambda_alias_A14F4FF8-allow-gateway-lambda-invoke_F4009ED6\\": {
        \\"action\\": \\"lambda:InvokeFunction\\",
        \\"function_name\\": \\"\${aws_lambda_alias.test-api-lambda_endpoint-lambda_alias_A14F4FF8.function_name}\\",
        \\"principal\\": \\"apigateway.amazonaws.com\\",
        \\"qualifier\\": \\"\${aws_lambda_alias.test-api-lambda_endpoint-lambda_alias_A14F4FF8.name}\\",
        \\"source_arn\\": \\"\${aws_api_gateway_rest_api.api-gateway-rest.execution_arn}/\${aws_api_gateway_stage.api-gateway-stage.stage_name}/\${aws_api_gateway_method.test-api-lambda_endpoint-method_B2F73D1A.http_method}\${aws_api_gateway_resource.test-api-lambda_endpoint_02B05EB1.path}\\"
      }
    },
    \\"aws_route53_record\\": {
      \\"test-api-lambda_api-gateway-certificate_certificate_record_BED7C8FC\\": {
        \\"depends_on\\": [
          \\"aws_acm_certificate.test-api-lambda_api-gateway-certificate_8D77B940\\"
        ],
        \\"name\\": \\"\${tolist(aws_acm_certificate.test-api-lambda_api-gateway-certificate_8D77B940.domain_validation_options)[0][\\\\\\"resource_record_name\\\\\\"]}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.test-api-lambda_api-gateway-certificate_8D77B940.domain_validation_options)[0][\\\\\\"resource_record_value\\\\\\"]}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.test-api-lambda_api-gateway-certificate_8D77B940.domain_validation_options)[0][\\\\\\"resource_record_type\\\\\\"]}\\",
        \\"zone_id\\": \\"\${aws_route53_zone.test-api-lambda_base-dns_subhosted_zone_B29688D6.zone_id}\\"
      },
      \\"test-api-lambda_apigateway-route53-domain-record_A8180A05\\": {
        \\"alias\\": [
          {
            \\"evaluate_target_health\\": true,
            \\"name\\": \\"\${aws_api_gateway_domain_name.test-api-lambda_api-gateway-domain-name_6DBA5B05.cloudfront_domain_name}\\",
            \\"zone_id\\": \\"\${aws_api_gateway_domain_name.test-api-lambda_api-gateway-domain-name_6DBA5B05.cloudfront_zone_id}\\"
          }
        ],
        \\"depends_on\\": [
          \\"aws_acm_certificate_validation.test-api-lambda_api-gateway-certificate_certificate_validation_6C09E593\\"
        ],
        \\"name\\": \\"\${aws_api_gateway_domain_name.test-api-lambda_api-gateway-domain-name_6DBA5B05.domain_name}\\",
        \\"type\\": \\"A\\",
        \\"zone_id\\": \\"\${aws_route53_zone.test-api-lambda_base-dns_subhosted_zone_B29688D6.zone_id}\\"
      },
      \\"test-api-lambda_base-dns_subhosted_zone_ns_B7A2A51A\\": {
        \\"name\\": \\"exampleapi.getpocket.dev\\",
        \\"records\\": \\"\${aws_route53_zone.test-api-lambda_base-dns_subhosted_zone_B29688D6.name_servers}\\",
        \\"ttl\\": 86400,
        \\"type\\": \\"NS\\",
        \\"zone_id\\": \\"\${data.aws_route53_zone.test-api-lambda_base-dns_main_hosted_zone_5F97294A.zone_id}\\"
      }
    },
    \\"aws_route53_zone\\": {
      \\"test-api-lambda_base-dns_subhosted_zone_B29688D6\\": {
        \\"name\\": \\"exampleapi.getpocket.dev\\"
      }
    },
    \\"aws_s3_bucket\\": {
      \\"test-api-lambda_endpoint-lambda_code-bucket_8153D419\\": {
        \\"acl\\": \\"private\\",
        \\"bucket\\": \\"pocket-lambda-endpoint\\",
        \\"force_destroy\\": true
      }
    },
    \\"aws_s3_bucket_public_access_block\\": {
      \\"test-api-lambda_endpoint-lambda_code-bucket-public-access-block_704E5627\\": {
        \\"block_public_acls\\": true,
        \\"block_public_policy\\": true,
        \\"bucket\\": \\"\${aws_s3_bucket.test-api-lambda_endpoint-lambda_code-bucket_8153D419.id}\\"
      }
    },
    \\"null_resource\\": {
      \\"test-api-lambda\\": {
      }
    }
  }
}"
`;
